File: binary-tree/binary-tree.js
class TreeNode {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

class BinaryTree {
  constructor() {
    this.root = null;
  }

  insert(value) {
    if (!this.root) {
      this.root = new TreeNode(value);
    } else {
      this._insertRecursive(this.root, value);
    }
  }

  _insertRecursive(node, value) {
    if (value < node.value) {
      if (node.left) {
        this._insertRecursive(node.left, value);
      } else {
        node.left = new TreeNode(value);
      }
    } else {
      if (node.right) {
        this._insertRecursive(node.right, value);
      } else {
        node.right = new TreeNode(value);
      }
    }
  }

  search(value) {
    return this._searchRecursive(this.root, value);
  }

  _searchRecursive(node, value) {
    if (!node) {
      return false;
    }
    if (node.value === value) {
      return true;
    }
    if (value < node.value) {
      return this._searchRecursive(node.left, value);
    }
    return this._searchRecursive(node.right, value);
  }

  dfs(value) {
    return this._dfsRecursive(this.root, value);
  }

  _dfsRecursive(node, value) {
    if (!node) {
      return false;
    }
    if (node.value === value) {
      return true;
    }
    if (this._dfsRecursive(node.left, value)) {
      return true;
    }
    if (this._dfsRecursive(node.right, value)) {
      return true;
    }
  }
}

const tree = new BinaryTree();
const valuesToInsert = [5, 3, 1, 10, 15, 7, 20];

valuesToInsert.forEach(value => tree.insert(value));

console.log(tree.search(5)); // true
console.log(tree.search(4)); // false
console.log(tree.search(10)); // true
console.log(tree.search(15)); // true
console.log("dfs: ", tree.dfs(20)); // DFS

File: binary-tree/binary-tree.py
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None


class BinaryTree:
    def __init__(self):
        self.root = None

    def insert(self, val):
        if not self.root:
            self.root = TreeNode(val)
        else:
            self._insert_recursive(self.root, val)

    def _insert_recursive(self, node, val):
        if val < node.val:
            if node.left:
                self._insert_recursive(node.left, val)
            else:
                node.left = TreeNode(val)
        else:
            if node.right:
                self._insert_recursive(node.right, val)
            else:
                node.right = TreeNode(val)

    def search(self, val):
        return self._search_recursive(self.root, val)

    def _search_recursive(self, node, val):
        if not node:
            return False
        if node.val == val:
            return True
        if val < node.val:
            return self._search_recursive(node.left, val)
        return self._search_recursive(node.right, val)

    def dfs(self, data):
        return self._dfs_recursive(self.root, data)

    def _dfs_recursive(self, node, data):
        if node:
            print(node.val)
        if node is None:
            return False
        if node.val == data:
            return True
        if self._dfs_recursive(node.left, data):
            return True
        if self._dfs_recursive(node.right, data):
            return True


tree = BinaryTree()
values_to_insert = [5, 3, 1, 10, 15, 7, 20]
for val in values_to_insert:
    tree.insert(val)

print(tree.search(5)) 
print(tree.search(4)) 
print(tree.search(10)) 
print(tree.search(15)) 
print(tree.dfs(20))